To add support for P construct (for | and + operations), and K construct

thomasyip opened the pull request:
Currently, `django-trusts` supports permission check with the following syntax:

```
from trusts.decorators import permission_required
from app.models import Xyz

@permission_required('app.change_xyz', fieldlookups_kwargs={'pk': 'xyz_id'})
def edit_xyz_view(request, xyz_id):
  # ...
  pass
```

And, the goal is to improve the syntax to allow for union ("|") and intersect ("+").

```
from trusts.decorators import permission_required, P, K
from app.models import Xyz

@permission_required(
      P('app.moderate_xyz', pk=K('xyz_id')) 
    | P('app.update_xyz:own', pk=K('xyz_id'))
)
def edit_xyz_view(request, xyz_id):
  # ...
  pass
```

Where `K` for `kwargs`, which works like combined `fieldlookups_kwargs`,  `fieldlookups_getparams` and `fieldlookups_postparams` in existing `permission_required` method. Those three param should still be supported, in case there is duplication among the lookups.

Where `P` should work somewhat similar to Q() in QuerySet. https://docs.djangoproject.com/es/1.9/topics/db/queries/#complex-lookups-with-q-objects

It should support Union `|` and Intersect `&` operators.

---

Moving repo for issue #2.

core2duo committed:
🆕 Added P class and tests

core2duo committed:
🔨 Fixed tests

core2duo committed:
💈 Moved permission logic from P class to @permission_required

thomasyip commented on trusts/decorators.py:
```diff
@@ -119,4 +194,31 @@ def check_perms(request, *args, **kwargs):
+            if fieldlookups is not None:
+                items = _get_permissible_items(request, p.perm, fieldlookups)
+                if items is None:
+                    p.permitted = False
```
Not sure it is a good idea to modify P instance. P is being reused for each http request.

core2duo replied:
When you use "P('app.moderate_xyz', pk=K('xyz_id')) | P('app.update_xyz:own', pk=K('xyz_id'))" you create a new P instance every time, so it doesn't matter will we modify it or not. Django can't cache this, because it has to call view_function every time (when url matches of course)

core2duo replied:
The only problem here it's if you want to use P class like this:

``` python
p1 = P('app.moderate_xyz', pk=K('xyz_id')) 
p2 = P('app.update_xyz:own', pk=K('xyz_id'))
p3 = P('app.delete_xyz:own', pk=K('xyz_id'))

@permission_required(p1 | p2)
def edit_xyz_view(request, xyz_id):
    pass

@permission_required(p2 & p3)
def edit_xyz_view(request, xyz_id):
    pass
```

But it's easy to solve, we just have to use copy() or deepcopy() in `__and__` and `__or__` methods.

thomasyip replied:
I think p instance should be "static". The main use is `@permission_required(P(...))`. The P constructor is called only once. The same P will serve any future HTTP request.

I think `permission_required` should call a method to walk the tree, and keep state within the method.

thomasyip commented on trusts/decorators.py:
```diff
@@ -119,4 +194,31 @@ def check_perms(request, *args, **kwargs):
+                    continue
+
+            if request.user.has_perms((p.perm, ), items):
+                p.permitted = True
```
My comment was really for this line.

core2duo committed:
💈 Made P objects "static" by keeping boolean flag.

thomasyip committed:
🛠 Refactored code to move permission checks out of P class. i

Refactored internal method params order to make them more consistence.

Cleaned up ununsed code. Converted to use keyword arguments for fieldlookups.
(lay a path for supporting K).

thomasyip committed:
🆕 Implemented decorators K(), G(), O() expression. Added tests.

thomasyip committed:
✏️ Updated doc for decorators: added P() expression, K(), G(), O() lookups description.

thomasyip committed:
🏷 Updated version number to 0.10.3.
